### ISSUE TITLE
Custom Location Downloads

### ISSUE THREAD

**Issue by snarkywolverine** [jmshrv/finamp#842]
Hello!

I see that there's some support in the android app for downloading to a custom location - though there's also a huge warning about it being buggy.

Are there steps I can take as a user to work around the permissions issues? I was hoping there'd be something documented (similar to https://github.com/jmshrv/finamp/issues/134#issuecomment-1376841753 ) to help me get this to work, even if there are some extra hoops to jump through.

Thanks!

**Comment by jmshrv** [comment]
The biggest thing you can do is switch to the redesign beta - the download system was completely overhauled and is now much more reliable. You can join the beta on Google Play :)

**Comment by Chaphasilor** [comment]
Also, did you apply that command you linked yet? Did it help?

**Comment by snarkywolverine** [comment]
Thanks for the quick reply - I'll give the beta a try.  The linked command above allowed me to download to the app data folder on my SD card - but I still couldn't read that file from elsewhere (which was my original goal).

**Comment by snarkywolverine** [comment]
Update - I'm on the beta, but still haven't cracked the code. I can download to the default directory just fine - both on the SD card and internal storage.   However, specifying a different directory results in failures.  In the logs, I see a TaskException "Operation not permitted" and an exception TaskHttpException, response code 416: Range Not Satisfiable when downloading <x>".

I haven't tried rerunning the above command yet...but since I can write to the appdir on the SD card, I suspect that's not the issue.

I looked at my Jellyfin server logs, expecting to see some details on the HTTP 416 ... but there's nothing relevant in that log.

Any other ideas I should try, or should I include more log information?

Samsung A32 5G, Android 13

**Comment by jmshrv** [comment]
hmm, could you send those Finamp logs? Custom paths in Android are still a massive pain so it wouldn't surprise me if the new system still isn't perfect (although in my mind the old system was practically broken for custom paths anyway)

**Comment by gozes** [comment]
So just checking, before I open another issue. is the adb shell commands still require even on the beta app, which I have not join yet, or should external path work on the stable version from the play store?

**Comment by Chaphasilor** [comment]
@gozes the download system on the beta is much more reliable than the stable version, so you should give it a try. you can join the beta on the Play Store.  
That being said, there shouldn't be a huge difference between custom location support in beta vs stable. You should be able to download to any path on your device which you can select using the system's file picker. The linked ADB command might help with accessing actual SD cards plugged into your device.  
There might be some issues with downloading track images in the beta, but the download of the audio files definitely works

**Comment by gozes** [comment]
Noted, will have a go at it and see how it goes. I know the ui redesign takes priority but any plans to at least move the download system to stable any time soon?

**Comment by Chaphasilor** [comment]
Not really. I don't want to push a half-finished redesign to the stable version, and untangling the download system from the redesign is basically impossible.  
What's the problem with using the beta? (honestly curious, maybe there's something we can improve in how the beta release is handled)

**Comment by gozes** [comment]
I don't have much of an issue with using the beta to be honest. I'm just trying to work out from looking at the open issues what the state of it is for my used case which is mostly offline on a DAP, hiby m300, with a micro SD card. At the moment I have the 1tb SD card in the m300 formatted as internal storage so that finamp can easily download the music to it without permission issues. So far that seems to be working fine but I have not tried it on long listings sessions to see if it has any issues. Given it's formatted that I'm not sure how easily it would be to pop the SD card on a computer and get the files from it assuming all my system backups were to fail and I couldn't recover any music from them. I know a bit of a big what-if but for that reason I would prefer if I could use the SD card as normal so that in the worst case I know I can pop it into my laptop and easily get the files in there back. I'll give the beta a try and see how it's goes but it would be good to know how stable it maybe for my used case
